fdbserver/LogSystem.cpp

```diff
@@ -369,6 +374,52 @@ bool LogPushData::writeTransactionInfo(int location, uint32_t subseq) {
 	return true;
 }
 
+void LogPushData::addTLogGroups(const std::vector<TLogGroupRef>& groups, Version commitVersion) {
+	for (const auto& group : groups) {
+		pGroupMessageBuilders->emplace(group->id(),
+		                               std::make_shared<ptxn::ProxySubsequencedMessageSerializer>(commitVersion));
+	}
+}
+
+std::map<ptxn::TLogGroupID, ptxn::SerializedTeamData> LogPushData::getGroupMutations(
+    const std::set<ptxn::TLogGroupID>& groups) {
+	std::map<ptxn::TLogGroupID, ptxn::SerializedTeamData> results;
+	for (const auto& [group, serializer] : *pGroupMessageBuilders) {
+		auto teamData = serializer->getAllSerialized();
+		results.emplace(group, teamData);
+	}
+	return results;
+}
+
+// TODO: we deserialize mutations sent from Resolvers and serialized to pGroupMessageBuilders.
+// It would be nice that ProxySubsequencedMessageSerializer can be set with
+// serialized data.
+void LogPushData::setGroupMutations(
+    const std::map<ptxn::TLogGroupID, std::unordered_map<ptxn::StorageTeamID, StringRef>>& groupMutations,
+    Version commitVersion) {
+	for (const auto& [group, teamData] : groupMutations) {
+		if (pGroupMessageBuilders->count(group) == 0) {
+			pGroupMessageBuilders->emplace(group,
+			                               std::make_shared<ptxn::ProxySubsequencedMessageSerializer>(commitVersion));
+		}
+		auto& writer = (*pGroupMessageBuilders)[group];
```

Comment:
consider:

auto it = pGroupMessageBuilders->find(group);
if (it == pGroupMessageBuilders->end()) {
  it = pGroupMessageBuilders->emplace(group,
			                               std::make_shared<ptxn::ProxySubsequencedMessageSerializer>(commitVersion)).first;
}


doing `(*pGroupMessageBuilders)[group];` always worries me that what if `group` doesn't exist. It's obvious not the case here since the creation is right above, but when more code is added in between, it becomes less readable.

Reply:
Good idea and it avoids look up `group` twice! Done.